libcxx/include/__tree

```diff
@@ -1787,26 +1988,47 @@ __tree<_Tp, _Compare, _Allocator>::__emplace_hint_unique_key_args(
 
 template <class _Tp, class _Compare, class _Allocator>
 template <class... _Args>
-typename __tree<_Tp, _Compare, _Allocator>::__node_holder
+_LIBCPP_CONSTEXPR_SINCE_CXX26 typename __tree<_Tp, _Compare, _Allocator>::__node_holder
 __tree<_Tp, _Compare, _Allocator>::__construct_node(_Args&&... __args) {
   __node_allocator& __na = __node_alloc();
   __node_holder __h(__node_traits::allocate(__na, 1), _Dp(__na));
+
+  #if _LIBCPP_STD_VER >= 26
```

Comment:
I think we should use something like
```C++
std::__construct_at(std::addressof(__h), __na, std::forward<_Args>(__args)...)
```
in all modes.

And use `__node_traits::construct` in the constructor of `__node`.

Reply:
> And use __node_traits::construct in the constructor of __node.

not sure I understand where this is supposed to be as:

https://github.com/llvm/llvm-project/blob/ea14834966ad666856246f24de46397cbfc932eb/libcxx/include/__tree#L754C34-L754C45 

is the definition of `__node`

Reply:
> > And use __node_traits::construct in the constructor of __node.
> 
> not sure I understand where this is supposed to be as:
> 
> https://github.com/llvm/llvm-project/blob/ea14834966ad666856246f24de46397cbfc932eb/libcxx/include/__tree#L754C34-L754C45
> 
> is the definition of `__node`

Hmm... `__node` is actually `__tree_node`, so we should modify that class. It seems necessary to wrap the `__value_` member into a `union`.

Reply:
> What would be the reason to do this?

We need to use allocator's `construct` function (if any) to construct the element.

> // what would be this part?

I think nothing else should be added to the union. We just need to call `__node_traits::construct(__a, std::addressof(__value_), __na, std::forward<_Args>(__args)...)` in the constructor.

Reply:
> > In all modes
> 
> Should this be submitted as a separate patch then? If we're also changing the existing runtime code path

I think so. But I overlooked C++03 mode before. Perhaps need to do different things (or keep old strategy?) in C++03 mode as some uses of `union` might be ill-formed.

Reply:
Let me know how you want to proceed.

As far as I can tell, in an early test, getting an allocator(the right type) to the `__tree_node` class may not be trivial and cause some duplication of the typedefs...

i.e.

```c++
template <class _Tp, class _VoidPtr>
class _LIBCPP_STANDALONE_DEBUG __tree_node : public __tree_node_base<_VoidPtr> {
public:
  using __node_value_type _LIBCPP_NODEBUG = __get_node_value_type_t<_Tp>;
  union {
    __node_value_type __value_;
  };
constexpr __tree(__node_allocator& __na, ...) {

__node_traits::construct(__na, ...);
}

};

```

`__node__allocator` and `__node_traits` are not trivial to get without changing the template args...

Reply:
> `__node__allocator` and `__node_traits` are not trivial to get without changing the template args...

Anyway, this constructor needs to be a template, so perhaps we should use

```C++
template <class _Alloc, class... _Args>
constexpr explicit __tree_node(_Alloc& __na, _Args&&... __args) {
  allocator_traits<_Alloc>::construct(__na, std::addressof(__value_), std::forward<_Args>(__args)...);
}
```

Reply:
Created: https://github.com/llvm/llvm-project/pull/153908